File: python/valid_palindrome_2.py
class Solution:
    def valid(self, s, left, right, chance):
        while left < right:
            if s[left] != s[right]:
                if chance is True:
                    return False

                ret_left_handled = self.valid(s, left + 1, right, True)
                ret_right_handled = self.valid(s, left, right - 1, True)

                return ret_left_handled or ret_right_handled

            left += 1
            right -= 1

        return True

    def validPalindrome(self, s: str) -> bool:
        return self.valid(s, 0, len(s) - 1, False)


s = Solution()
print(s.validPalindrome("aba"))
print(s.validPalindrome("abca"))
print(s.validPalindrome("abc"))
print(s.validPalindrome("a"))
print(s.validPalindrome("aa"))

File: scala/ValidPalindrome2.scala
import scala.collection.mutable

object ValidPalindrome2 extends App {
  def validPalindrome(s: String): Boolean = {
    val (l, r) = (0, s.length - 1)
    val q = mutable.Stack[(Int, Int)]()
    q.addOne((l, r))
    var chance = true
    while(q.nonEmpty) {
      val (left, right) = q.pop()
      if(left >= right) return true
      if(s(left) == s(right)) {
        q.addOne((left + 1, right - 1))
      } else if(chance) {
        chance = false
        q.addOne((left, right - 1))
        q.addOne((left + 1, right))
      }
    }
    false
  }

  println(validPalindrome("aba"))
  println(validPalindrome("abca"))
  println(validPalindrome("abc"))
  println(validPalindrome("dabcad"))
  println(validPalindrome("afkfga"))
}
